### PR TITLE
sql: properly clean up the KV txn after various SQL error

### PR DESCRIPTION
**Pull request by andreimatei**
I'm gonna write a test for the bug too, but it's a bit tricky.

@bdarnell or @tschottdorf please look at the first commit if you can and see what you think.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5586)

### CODE REVIEW

**Comment by RaduBerinde** [comment]
Commit 2 LGTM.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-:-KDk3EGsL0RRzWb7z-nO:-502661496)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[client/txn.go, line 130 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDkUgjW01yCfhbNHozn:-KDkUgjW01yCfhbNHozo:-735007632) ([raw file](https://github.com/cockroachdb/cockroach/blob/567f4c54de1c2b20038ec54461a71e64dcd2b501/client/txn.go#L130)):*
How about calling this IsPending and inverting the logic? That way it's more explicit.

---

*[client/txn.go, line 432 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDkRi5gp2r5fh6uyO18:-KDkRi5gp2r5fh6uyO19:1203399939) ([raw file](https://github.com/cockroachdb/cockroach/blob/567f4c54de1c2b20038ec54461a71e64dcd2b501/client/txn.go#L432)):*
Can you make this just like in Commit where you check the state before exiting, instead of putting it in the else clause

---

*[sql/executor.go, line 754 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDkTv3lN4T2m7Jhtg-K:-KDkTv3lN4T2m7Jhtg-L:-255567376) ([raw file](https://github.com/cockroachdb/cockroach/blob/567f4c54de1c2b20038ec54461a71e64dcd2b501/sql/executor.go#L754)):*
what wrong with using pErr here?

---

*[sql/session.go, line 197 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDkSiTFihQvXYRYy2yf:-KDkSiTFihQvXYRYy2yg:1630543470) ([raw file](https://github.com/cockroachdb/cockroach/blob/567f4c54de1c2b20038ec54461a71e64dcd2b501/sql/session.go#L197)):*
Its a pity you need to export IsFinalized because of a panic here. Might as well just use ts.txn.Proto.Status == roachpb.PENDING

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586)*

File: client/txn.go [hunk 1]
```diff
@@ -122,6 +122,14 @@ func NewTxn(ctx context.Context, db DB) *Txn {
 	return txn
 }
 
+// IsFinalized returns true if this Txn has been committed or rolled back and
+// should therefore not be used for any more KV operation.
```
Comment (spencerkimball) [inline client/txn.go]: s/operation/operations/

File: client/txn.go [hunk 2]
```diff
@@ -407,8 +419,20 @@ func (txn *Txn) CommitBy(deadline roachpb.Timestamp) *roachpb.Error {
 }
 
 // Rollback sends an EndTransactionRequest with Commit=false.
+// The txn's status is set to ABORTED in case of error. txn is
+// considered finalized and cannot be used to send any more commands.
 func (txn *Txn) Rollback() *roachpb.Error {
-	return txn.sendEndTxnReq(false /* commit */, nil)
+	err := txn.sendEndTxnReq(false /* commit */, nil)
+	if err != nil {
+		// TODO(andrei): I'm not completely sure setting the status here is required;
```
Comment (spencerkimball) [inline client/txn.go]: `txn.Proto.Status = ABORTED` is not set on the `send()` return path unless the error experienced was `TransactionAbortedError` (`txn.Proto` is only updated on abort and retry errors). It seems reasonable to set this here. Remove the TODO and add a comment mentioning the desire to poison the txn on a failure to rollback.

**Comment by bdarnell** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[client/txn.go, line 429 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDssu5AhefWUei7qP-3:-KDssu5AhefWUei7qP-4:-1883114953) ([raw file](https://github.com/cockroachdb/cockroach/blob/77bda86c0164c6e46cb6a8fab2e885f98839e356/client/txn.go#L429)):*
This makes me nervous. I think it would be fine to set status to `ABORTED` here, but doing so deeper in the stack would be bad (e.g. if an `ABORTED` transaction proto made it into the `intentResolver` while the on-disk record was still `PENDING`, it would start removing intents that could potentially become committed (i can imagine some scenarios involving network errors and out-of-order replays that might cause this). 

Instead, how about leaving the `Status` field to the server's control and adding an explicit finalized boolean field to `client.Txn` to be set when `Commit` or `Rollback` is called?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586)*

**Comment by tschottdorf** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[client/txn.go, line 429 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDssu5AhefWUei7qP-3:-KDyFdXJl-zzRhD0f4O_:-1662576751) ([raw file](https://github.com/cockroachdb/cockroach/blob/77bda86c0164c6e46cb6a8fab2e885f98839e356/client/txn.go#L429)):*
I'd also prefer this. Fairly sure it's possible to keep using the `*Txn` (well, that shouldn't be possible, but that's a different story), and it's cleaner to  add a boolean `finalized` which is checked in `(*txnSender).Send()`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586)*

**Comment by andreimatei** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[client/txn.go, line 429 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDssu5AhefWUei7qP-3:-KDyFUTz5bbQzbG5Ol1W:1225517724) ([raw file](https://github.com/cockroachdb/cockroach/blob/77bda86c0164c6e46cb6a8fab2e885f98839e356/client/txn.go#L429)):*
ok, separate field.

---

*[client/txn.go, line 126 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDjywOJIDUizcgh6qHD-r2-126:-KDyfI7mYwTCA_ZtDV51:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/567f4c54de1c2b20038ec54461a71e64dcd2b501/client/txn.go#L126)):*
Done.

---

*[client/txn.go, line 130 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDkUgjW01yCfhbNHozn:-KDyfJDGI-e-IXTxdMfF:952454827) ([raw file](https://github.com/cockroachdb/cockroach/blob/567f4c54de1c2b20038ec54461a71e64dcd2b501/client/txn.go#L130)):*
the code now uses a dedicated field, so the objection would probably go away.

Also who knows what other states we add to this tx. For example there's another PR out  adding a retrying  field that could be a state; we might also need a state to differentiate between a txn that has already been used to send some commands vs a fresh one. `IsPending()` can become less clearly defined than it'd be today. `IsFinalized()` sounds nice and stable.

---

*[client/txn.go, line 432 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDkRi5gp2r5fh6uyO18:-KDyfAoQYAPluhH8PCSP:166904447) ([raw file](https://github.com/cockroachdb/cockroach/blob/567f4c54de1c2b20038ec54461a71e64dcd2b501/client/txn.go#L432)):*
this code is gone

---

*[sql/executor.go, line 754 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDkTv3lN4T2m7Jhtg-K:-KDygawIoOIu5ehiFUhW:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/567f4c54de1c2b20038ec54461a71e64dcd2b501/sql/executor.go#L754)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586)*

**Comment by tschottdorf** [comment]
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 129 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDytc6P0F0bPbeOVdAB:-KDytc6P0F0bPbeOVdAC:-480130859) ([raw file](https://github.com/cockroachdb/cockroach/blob/aac8941b67f5f30f6459e70c4db757c812b8d823/client/txn.go#L129)):*
Add more precise semantics to the comment. For example, it's not clear that `IsFinalized()` returns false after an unsuccessful attempt to commit.

---

*[client/txn.go, line 387 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDyu8HqzH_CYtWZFkNC:-KDyu8HqzH_CYtWZFkND:-1539303185) ([raw file](https://github.com/cockroachdb/cockroach/blob/aac8941b67f5f30f6459e70c4db757c812b8d823/client/txn.go#L387)):*
> If the command completes successfully, the txn is considered finalized. On error, no attempt is made to clean up the (possibly still pending) transaction.

---

*[client/txn.go, line 411 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDyuPiR8JL6fMWUdbWo:-KDyuPiR8JL6fMWUdbWp:-359697853) ([raw file](https://github.com/cockroachdb/cockroach/blob/aac8941b67f5f30f6459e70c4db757c812b8d823/client/txn.go#L411)):*
this doesn't seem right, it's perfectly fine to call Commit() multiple times in the `db.Txn` closure (in fact, it's expected).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586)*

**Comment by andreimatei** [comment]
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[client/txn.go, line 129 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDytc6P0F0bPbeOVdAB:-KDyxH5AL3E-Y3P_3s2x:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aac8941b67f5f30f6459e70c4db757c812b8d823/client/txn.go#L129)):*
Done.

---

*[client/txn.go, line 387 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDyu8HqzH_CYtWZFkNC:-KDyxQJkWhfeH5ZJkvez:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aac8941b67f5f30f6459e70c4db757c812b8d823/client/txn.go#L387)):*
Done.

---

*[client/txn.go, line 411 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDyuPiR8JL6fMWUdbWo:-KDywSly2dlyHieDDXPj:1041374127) ([raw file](https://github.com/cockroachdb/cockroach/blob/aac8941b67f5f30f6459e70c4db757c812b8d823/client/txn.go#L411)):*
there's `CommitNoCleanup` for that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r1, 1 of 3 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[sql/executor.go, line 753 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDyycI0_Wv8BcQ2dfDW:-KDyycI1NTslc1dVYoF1:521263655) ([raw file](https://github.com/cockroachdb/cockroach/blob/629125f211b359db1806b8eaa54bf6375aa1c1cd/sql/executor.go#L753)):*
this is a weird style, just pErr := ... and put the `if` on the next line.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586)*

File: client/txn.go [hunk 3]
```diff
@@ -371,6 +387,9 @@ func (txn *Txn) CommitNoCleanup() *roachpb.Error {
 // optional, but more efficient than relying on the implicit commit
 // performed when the transaction function returns without error.
 // The batch must be created by this transaction.
+// If the command completes successfully, the txn is considered finalized. On
+// error, no attempt is made to clean up the (possibly still pending)
+// transaction.
```
Comment (tschottdorf) [inline client/txn.go]: It's pretty inconsistent that `CommitInBatch` behaves differently from `Commit` with respect to cleanup. I think `CommitNoCleanup` ought to be renamed to `Commit` and that the current `Commit` should be `CommitOrRollback` (or your favorite name for attempt commit, attempt rollback if the commit fails).

**Comment by tschottdorf** [comment]
LGTM with some comments.

**Comment by andreimatei** [comment]
Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[client/txn.go, line 392 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDjywOJIDUizcgh6qHD-r5-392:-KDz2NiNb9jAo2aiRzkw:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d22c5dc02dfcc4d97ba9f276bc522df3eba6e2/client/txn.go#L392)):*
Done.

---

*[sql/executor.go, line 753 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586#-KDyycI0_Wv8BcQ2dfDW:-KDz2M901uf1xdyyqbXf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/629125f211b359db1806b8eaa54bf6375aa1c1cd/sql/executor.go#L753)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5586)*

**Pull request merged**